extra/lib/plausible/funnels.ex

```diff
@@ -92,4 +108,25 @@ defmodule Plausible.Funnels do
 
     Repo.one(q)
   end
+
+  @spec for_site_query(Plausible.Site.t(), Keyword.t()) :: Ecto.Query.t()
+  def for_site_query(site, opts \\ []) do
+    if opts[:preload_goals?] do
+      from(f in Funnel,
+        inner_join: steps in assoc(f, :steps),
+        where: f.site_id == ^site.id,
+        group_by: f.id,
+        order_by: [desc: :id],
+        preload: [steps: :goal]
```

Comment:
```suggestion
        preload: [steps: {steps, :goal}]
```

or even go full query-based preload and do:

```elixir
inner_join: goal in assoc(steps, :goal),
...
preload: [steps: {steps, goal: goal}]
```

Otherwise, the `steps` join remains completely unused, as both steps and goals are preloaded with separate queries.

Reply:
um I'm afraid every suggestion breaks pagination (test/plausible_web/plugins/api/controllers/funnels_test.exs), will revisit in the morning